### DISCUSSION TITLE
Nonlinear Optics Simulation

### DISCUSSION THREAD

**Discussion by jiecaix** [BLAST-WarpX/warpx#4923] [Q&A] [answered]
I noticed that Warpx could setting a set of "macroscopic.sigma_function(x,y,z), macroscopic.epsilon_function(x,y,z) and/or macroscopic.mu_function(x,y,z)" functions in "em_solver_medium" module.
And i wonder if it possible to use this function to set a nonlinear optics simulation, such as what a FDTD simulation software done like "MEEP". For example, those function could be influenced by not only position but the laser intensity and the interaction with medium.

**Comment by ax3l** [comment, marked as answer]
Yes, that would definitely work.

I think we do not yet have support to pass a complex function to these local material properties, but otherwise this should be easy to add to WarpX.
For instance, one could add extra inputs for the imaginary part and update the macroscopic solvers slightly.

For injecting a light source, our current approach using an antenna should work for this as well.
